src/cloudformation_cli_python_lib/recast.py

```diff
@@ -0,0 +1,131 @@
+    for k, v in json_data.items():
+        if isinstance(v, dict):
+            child_cls = _field_to_type(cls.__dataclass_fields__[k].type, k, classes)
+            recast_object(child_cls, v, classes)
+        elif isinstance(v, list):
+            json_data[k] = _recast_lists(cls, k, v, classes)
+        elif isinstance(v, set):
+            json_data[k] = _recast_sets(cls, k, v, classes)
+        elif isinstance(v, str):
+            dest_type = _field_to_type(cls.__dataclass_fields__[k].type, k, classes)
+            json_data[k] = _recast_primitive(dest_type, k, v)
+        else:
+            raise InvalidRequest(f"Unsupported type: {type(v)} for {k}")
+
+
+def _recast_lists(cls: Any, k: str, v: List[Any], classes: Dict[str, Any]) -> List[Any]:
+    # Leave as is if type is Any
+    if cls == typing.Any:
+        return v
+    casted_list: List[Any] = []
+    if "__dataclass_fields__" not in dir(cls):
+        pass
+    elif k in cls.__dataclass_fields__:
+        cls = _field_to_type(cls.__dataclass_fields__[k].type, k, classes)
+    for item in v:
+        casted_list.append(cast_sequence_item(cls, k, item, classes))
+    return casted_list
+
+
+def _recast_sets(cls: Any, k: str, v: Set[Any], classes: Dict[str, Any]) -> Set[Any]:
+    casted_set: Set[Any] = set()
+    if "__dataclass_fields__" in dir(cls):
+        cls = _field_to_type(cls.__dataclass_fields__[k].type, k, classes)
+    for item in v:
+        casted_set.add(cast_sequence_item(cls, k, item, classes))
+    return casted_set
+
+
+def cast_sequence_item(cls: Any, k: str, item: Any, classes: Dict[str, Any]) -> Any:
+    if isinstance(item, str):
+        return _recast_primitive(cls, k, item)
+    if isinstance(item, list):
+        return _recast_lists(cls, k, item, classes)
+    if isinstance(item, set):
+        return _recast_sets(cls, k, item, classes)
+    if isinstance(item, dict):
+        recast_object(cls, item, classes)
+        return item
+    raise InvalidRequest(f"Unsupported type: {type(item)} for {k}")
+
+
+def _recast_primitive(cls: Any, k: str, v: str) -> Any:
+    if cls == typing.Any:
+        # If the type is Any, we cannot guess what the original type was, so we leave
+        # it as a string
+        return v
+    if cls == bool:
+        if v.lower() == "true":
+            return True
+        if v.lower() == "false":
+            return False
+        raise InvalidRequest(f'value for {k} "{v}" is not boolean')
+    return cls(v)
+
+
+# yes, introspecting type hints is ugly, but hopefully only needed temporarily
+def _field_to_type(field: Any, key: str, classes: Dict[str, Any]) -> Any:  # noqa: C901
+    if field in [int, float, str, bool, typing.Any]:
+        return field
+    # If it's a ForwardRef we need to find base type
+    if isinstance(field, get_forward_ref_type()):
+        # Assuming codegen added an _ as a prefix, removing it and then getting the
+        # class from model classes
+        return classes[field.__forward_arg__[1:]]
```

Comment:
what's this actually returning?

Reply:
To prevent collisions when generating the model we sometimes point to it with an `_` prefix, and potentially it's a string of the class name, not the class itself ([example](https://github.com/jaymccon/cloudformation-cli-python-plugin/blob/nested_schema_and_recast/tests/lib/sample_model.py#L46-L48)) this returns the class that the string refers to.